### PR TITLE
Day 3: "Improvements to Speculation Rules" / Barry Pollard

### PR DESCRIPTION
**Pull request by tunetheweb**
As promised (even if a little late — sorry!)

Happy to take any suggestions or feedback here, over email or whereever!

### CODE REVIEW

File: hell/adventcalendar/2025/3/index.md [hunk 1]
```diff
@@ -1,35 +1,123 @@
...
 image: "advent25_3"
 ---
-Some text.
-Some text.
 
-Some text. Some text.
+The [Speculation Rules API](https://developer.mozilla.org/docs/Web/API/Speculation_Rules_API) allows you to speed up future navigations by prefetching or even prerendering URLS in advance of a user actually clicking a link. When the link is clicked, the speculation is used and the user experiences a faster load than if no speculation was used.
 
-## Heading
+[https://schepp.dev/](Schepp) covered [the API in last year's post](/adventcalendar/2024/28/) and discussed the `eagerness` value which allowed you to, for example, hover over a link to speculatively prerender it with the simple addition of this rule to your HTML:
```
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]:
```suggestion
[Schepp](https://schepp.dev/) covered [the API in last year's post](/adventcalendar/2024/28/) and discussed the `eagerness` value which allowed you to, for example, hover over a link to speculatively prerender it with the simple addition of this rule to your HTML:
```

File: hell/adventcalendar/2025/3/index.md [hunk 2]
```diff
@@ -1,35 +1,123 @@
...
+
+<p class="highlight"><strong>Note:</strong> Speculating has a cost. To both your users, and for sites with potential increased traffic and and resulting infrastructure usage. Always weigh those costs with the benefit to the user.</p>
+
+To conserve memory Chrome keeps up to two speculations in memory at a time. As the users scrolls further and new links — which are more likely to be click on — enter the viewport the old prerenders are cancelled. Should these links be re-speculated (for example, but the user scroll back up), then they will be fetched from the HTTP cache and so prerender even faster.
 
-## Heading
+## `eager` eagerness improvements
 
-Some text.
+As well as the improvements to the `moderate` eagerness, the `eager` value has also been recently changed to offer an option somewhere between `immediate` (where loinks are speculated as soon as possible) and `moderate`.
```
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]:
```suggestion
As well as the improvements to the `moderate` eagerness, the `eager` value has also been recently changed to offer an option somewhere between `immediate` (where links are speculated as soon as possible) and `moderate`.
```

File: hell/adventcalendar/2025/3/index.md [hunk 3]
```diff
@@ -1,35 +1,123 @@
...
+  }]
+}
+</script>
+```
+
+This hover technique is common functionality offered by libraries, plugins and frameworks but now this is baked right into the browser. Chromium-based browsers only for now but it is being worked on by Safari and Firefox, and as a progressive enhancement nothing is broken on non-supporting browsers — they just don't benefit from the performance improvement.
+
+Using mouse hover as a signal works well on desktop, but what about mobile where "hover" is not a really a thing? At least in browser terms — I'm sure many of us have hovered over a link with out fingers but as that is not detected by the browser as an event, we can't act upon it.
+
+Well solving that is just one of many improvements the API has seen in the last year that we'll cover in this post.
```
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]:
```suggestion
Well, solving that is just one of many improvements the API has seen in the last year that we'll cover in this post.
```

File: hell/adventcalendar/2025/3/index.md [hunk 4]
```diff
@@ -1,35 +1,123 @@
...
+  if (!mediaQuery.matches) {
+      document.querySelector('video').autoplay = true;
+  }
+</script>
+
+In the video you can see as you scroll down the page, links are being successfully prerendered ready for the user to browse to the page.
+
+<p class="highlight"><strong>Note:</strong> Speculating has a cost. To both your users, and for sites with potential increased traffic and and resulting infrastructure usage. Always weigh those costs with the benefit to the user.</p>
+
+To conserve memory Chrome keeps up to two speculations in memory at a time. As the users scrolls further and new links — which are more likely to be click on — enter the viewport the old prerenders are cancelled. Should these links be re-speculated (for example, but the user scroll back up), then they will be fetched from the HTTP cache and so prerender even faster.
```
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]:
```suggestion
To conserve memory, Chrome keeps up to two speculations in memory at a time. As the user scrolls further and new links — which are more likely to be clicked on — enter the viewport, the old prerenders are cancelled. Should these links be re-speculated (for example, but the user scrolls back up), then they will be fetched from the HTTP cache and so prerender even faster.
```

File: hell/adventcalendar/2025/3/index.md [hunk 5]
```diff
@@ -1,35 +1,123 @@
...
+<script>
+  const mediaQuery = window.matchMedia('(prefers-reduced-motion: reduce)');
+  if (!mediaQuery.matches) {
+      document.querySelector('video').autoplay = true;
+  }
+</script>
+
+In the video you can see as you scroll down the page, links are being successfully prerendered ready for the user to browse to the page.
+
+<p class="highlight"><strong>Note:</strong> Speculating has a cost. To both your users, and for sites with potential increased traffic and and resulting infrastructure usage. Always weigh those costs with the benefit to the user.</p>
```
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]: Now I'm wondering how dangerous it is. Can we take the first code snippet from your post, add it to our HTML boilerplate, and use it on every site?
How do we determine if it's over-speculating, and how do we deal with it?

File: hell/adventcalendar/2025/3/index.md [hunk 6]
```diff
@@ -1,35 +1,123 @@
...
-## Heading
+## `eager` eagerness improvements
 
-Some text.
+As well as the improvements to the `moderate` eagerness, the `eager` value has also been recently changed to offer an option somewhere between `immediate` (where loinks are speculated as soon as possible) and `moderate`.
 
+For desktop, `moderate` rules will trigger after a 10 millisecond hover, and for mobile we consider ALL links in the viewport after the user has stopped scrolling for 100 milliseconds rather than the more restrictive set of heuristics above.
 
-<p class="highlight"><strong>Note:</strong> Some text.</p>
+One common technique is to prefetch with `eager` and then upgrade to a `prerender` with `moderate` when you have more signals the users may click on the link:
```
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]: I don't understand what's going on here. I think you need to explain the difference between prefetch and prerender, because I would've assumed prerender implies prefetch. Also, it would help if you could describe what the browser does with this setting to remind us what eager and moderate do.

Adding to one my revious questions: is this something I would want to add to my HTML boilerplate?

File: hell/adventcalendar/2025/3/index.md [hunk 7]
```diff
@@ -1,35 +1,123 @@
...
+  }]
+}
+</script>
 ```
+
+In addition you can also restrict the links considered for speculating using the `where` object.
+
+## Further improvements to the API
+
+I'll close out with a sneak peak into the future as one further improvement being worked on is a middle group between the `prefetch` and `prerender`. This is useful for those sites concerned with any negative impact with fully prerendering a page.
```
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]:
```suggestion
I'll close out with a sneak peek into the future, as one further improvement being worked on is a middle group between the `prefetch` and `prerender`. This is useful for those sites concerned with any negative impact with fully prerendering a page.
```

File: hell/adventcalendar/2025/3/index.md [hunk 8]
```diff
@@ -1,35 +1,123 @@
...
+    url: "https://mastodon.social/@tunetheweb"
+    link_label: "@tunetheweb@mastodon.social"
+intro: "<p>The Speculation Rules API is becoming even better with improved mobile support, and a few other improvements to help speed up your websites.</p>"
 image: "advent25_3"
 ---
-Some text.
-Some text.
 
-Some text. Some text.
+The [Speculation Rules API](https://developer.mozilla.org/docs/Web/API/Speculation_Rules_API) allows you to speed up future navigations by prefetching or even prerendering URLS in advance of a user actually clicking a link. When the link is clicked, the speculation is used and the user experiences a faster load than if no speculation was used.
```
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]:
```suggestion
The [Speculation Rules API](https://developer.mozilla.org/docs/Web/API/Speculation_Rules_API) allows you to speed up future navigations by prefetching or even prerendering URLS in advance of a user actually clicking a link. When the link is clicked, the speculation is used, and the user experiences a faster load than if no speculation was used.
```

File: hell/adventcalendar/2025/3/index.md [hunk 9]
```diff
@@ -1,35 +1,123 @@
...
+{
+  "prerender": [{
+    "where": { "href_matches": "/*" },
+    "eagerness": "moderate"
+  }]
+}
+</script>
+```
+
+This hover technique is common functionality offered by libraries, plugins and frameworks but now this is baked right into the browser. Chromium-based browsers only for now but it is being worked on by Safari and Firefox, and as a progressive enhancement nothing is broken on non-supporting browsers — they just don't benefit from the performance improvement.
```
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]:
```suggestion
This hover technique is a common functionality offered by libraries, plugins, and frameworks, but now this is baked right into the browser. Chromium-based browsers only for now, but it is being worked on by Safari and Firefox, and as a progressive enhancement, nothing is broken on non-supporting browsers — they just don't benefit from the performance improvement.
```

File: hell/adventcalendar/2025/3/index.md [hunk 10]
```diff
@@ -1,35 +1,123 @@
...
+
+Using mouse hover as a signal works well on desktop, but what about mobile where "hover" is not a really a thing? At least in browser terms — I'm sure many of us have hovered over a link with out fingers but as that is not detected by the browser as an event, we can't act upon it.
```
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]:
```suggestion
Using mouse hover as a signal works well on desktop, but what about mobile, where "hover" is not really a thing? At least in browser terms — I'm sure many of us have hovered over a link without fingers, but as that is not detected by the browser as an event, we can't act upon it.
```

File: hell/adventcalendar/2025/3/index.md [hunk 11]
```diff
@@ -1,35 +1,123 @@
...
+
+## Improved mobile viewport heuristics
+
+Since hover does not exist, we have updated the `eagerness: moderate` on mobile to instead look at the following heuristics:
+
+- Anchor links within 30% vertical distance from the previous pointer down.
+- Anchor links at least 0.5x as big as the largest anchor in the viewport.
+- The browser waits 500 milliseconds after the user stopped scrolling.
+
+The aim of this is to try to avoid, as much as possible, over-speculating links users are less likely to click on. For example, because they are still scrolling, or they are very small links (terms and conditions and the like!).
```
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]:
```suggestion
The aim is to avoid, as much as possible, over-speculating links that users are less likely to click. For example, because they are still scrolling, or they are very small links (terms and conditions and the like!).
```

File: hell/adventcalendar/2025/3/index.md [hunk 12]
```diff
@@ -1,35 +1,123 @@
...
+  <figcaption><p>An example of scrolling down a mobile page triggering speculations.</p></figcaption>
+</figure>
+<script>
+  const mediaQuery = window.matchMedia('(prefers-reduced-motion: reduce)');
+  if (!mediaQuery.matches) {
+      document.querySelector('video').autoplay = true;
+  }
+</script>
+
+In the video you can see as you scroll down the page, links are being successfully prerendered ready for the user to browse to the page.
```
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]:
```suggestion
In the video, you can see that as you scroll down the page, links are being successfully prerendered, ready for the user to browse to the page.
```

File: hell/adventcalendar/2025/3/index.md [hunk 13]
```diff
@@ -1,35 +1,123 @@
...
+</script>
 ```
+
+In addition you can also restrict the links considered for speculating using the `where` object.
+
+## Further improvements to the API
+
+I'll close out with a sneak peak into the future as one further improvement being worked on is a middle group between the `prefetch` and `prerender`. This is useful for those sites concerned with any negative impact with fully prerendering a page.
+
+A new **Prender Until Script** option is currently available behind a flag in Chrome (`chrome://flags/#prerender-until-script`). Like it's name suggests, will start prerendering a page but then pause when it encounters a synchronous `<script>` element. Scripts with the `async` or `defer` attribute (or `module` scripts which are `defer` by default) will be downloaded but not executed.
```
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]:
```suggestion
A new **Prender Until Script** option is currently available behind a flag in Chrome (`chrome://flags/#prerender-until-script`). As its name suggests, it will start prerendering a page, but pause when it encounters a synchronous `<script>` element. Scripts with the `async` or `defer` attribute (or `module` scripts which are `defer` by default) will be downloaded but not executed.
```

File: hell/adventcalendar/2025/3/index.md [hunk 14]
```diff
@@ -1,35 +1,123 @@
...
+
+This means:
+
+- Pages without any JavaScript can be fully prerendered.
+- Pages with only async/deferred JavaScript can be fully prerendered, with JavaScript executed on navigation.
+- Pages with only sync `<script>` JavaScript can start prerender but pause before any `<script>` causes any intended consequences. They will continue to download subresources (thanks to the [preload scanner](https://web.dev/articles/preload-scanner)) so still have a big performance benefit over `prefetch`.
```
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]:
```suggestion
- Pages with only sync `<script>` JavaScript can start prerender, but pause before any `<script>` causes any intended consequences. They will continue to download subresources (thanks to the [preload scanner](https://web.dev/articles/preload-scanner)), so they still have a significant performance benefit over `prefetch`.
```

File: hell/adventcalendar/2025/3/index.md [hunk 15]
```diff
@@ -1,35 +1,123 @@
...
+
+After enabling the flag, you can use this new mode in exactly the same way as `preftech` and `prerender`:
```
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]:
```suggestion
After enabling the flag, you can use this new mode in exactly the same way as `prefetch` and `prerender`:
```

File: hell/adventcalendar/2025/3/index.md [hunk 16]
```diff
@@ -1,35 +1,123 @@
...
+{
+  "prerender_until_script": [{
+    "where": { "href_matches": "/*" },
+    "eagerness": "moderate"
+  }]
+}
+</script>
+```
+
+This enhancement should be released next year, but in the mean time have a place and <a href="https://bsky.app/profile/tunetheweb.com">let me know</a> how you get on!
```
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]:
```suggestion
This enhancement should be released next year, but in the meantime, have a place and <a href="https://bsky.app/profile/tunetheweb.com">let me know</a> how you get on! 
```

File: hell/adventcalendar/2025/3/index.md [hunk 17]
```diff
@@ -1,35 +1,123 @@
...
+  }]
+}
+</script>
+```
+
+This enhancement should be released next year, but in the mean time have a place and <a href="https://bsky.app/profile/tunetheweb.com">let me know</a> how you get on! 
+
+## Conclusion
+
+The Speculation Rules API continues to improve with new options and features to help site owners deliver fast, HTML-driven, websites! With our first signs of cross-browser adoption of the API, I'd speculate (boom! boom!) 2026 will be another bumper year so the API — and for users of sites that implement it.
```
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]:
```suggestion
The Speculation Rules API continues to improve with new options and features to help site owners deliver fast, HTML-driven websites! With our first signs of cross-browser adoption of the API, I'd speculate (boom! boom!) 2026 will be another bumper year for the API and for users of sites that implement it.
```

File: hell/adventcalendar/2025/3/index.md [hunk 18]
```diff
@@ -1,35 +1,123 @@
...
+Using mouse hover as a signal works well on desktop, but what about mobile where "hover" is not a really a thing? At least in browser terms — I'm sure many of us have hovered over a link with out fingers but as that is not detected by the browser as an event, we can't act upon it.
+
+Well solving that is just one of many improvements the API has seen in the last year that we'll cover in this post.
+
+## Improved mobile viewport heuristics
+
+Since hover does not exist, we have updated the `eagerness: moderate` on mobile to instead look at the following heuristics:
+
+- Anchor links within 30% vertical distance from the previous pointer down.
+- Anchor links at least 0.5x as big as the largest anchor in the viewport.
```
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]:
```suggestion
- Anchor links at least 0.5× as big as the largest anchor in the viewport.
```

File: hell/adventcalendar/2025/3/index.md (hunk 5 above)
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]:
```suggestion
<p class="highlight"><strong>Note:</strong> Speculating has a cost. To both your users and for sites with potential increased traffic and resulting infrastructure usage. Always weigh those costs against the benefit to the user.</p>
```

File: hell/adventcalendar/2025/3/index.md [hunk 19]
```diff
@@ -1,35 +1,123 @@
...
+
+To conserve memory Chrome keeps up to two speculations in memory at a time. As the users scrolls further and new links — which are more likely to be click on — enter the viewport the old prerenders are cancelled. Should these links be re-speculated (for example, but the user scroll back up), then they will be fetched from the HTTP cache and so prerender even faster.
 
-## Heading
+## `eager` eagerness improvements
 
-Some text.
+As well as the improvements to the `moderate` eagerness, the `eager` value has also been recently changed to offer an option somewhere between `immediate` (where loinks are speculated as soon as possible) and `moderate`.
 
+For desktop, `moderate` rules will trigger after a 10 millisecond hover, and for mobile we consider ALL links in the viewport after the user has stopped scrolling for 100 milliseconds rather than the more restrictive set of heuristics above.
```
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]:
```suggestion
On desktop, `moderate` rules trigger after a 10 millisecond hover, and on mobile we consider ALL links in the viewport after the user has stopped scrolling for 100 milliseconds, rather than the more restrictive set of heuristics above.
```

File: hell/adventcalendar/2025/3/index.md (hunk 6 above)
Comment (matuzo) [inline hell/adventcalendar/2025/3/index.md]:
```suggestion
One common technique is to prefetch with `eager` and then upgrade to a `prerender` with `moderate` when you have more signals that the users may click on the link:
```

**Review by matuzo — COMMENTED** [review]
Great stuff, thank you. The post is clear but it's not absolutely clear how to use the API efficiently. I've added some questions.

File: hell/adventcalendar/2025/3/index.md (hunk 5 above)
Comment (tunetheweb) [reply hell/adventcalendar/2025/3/index.md]: Speculation has a number of risks or "dangers":
1. Breaking things. For example speculating a "logout" link could log you out. Or an "add to cart" could update a card. And if the user doesn't visit those pages then that wouldn't be a good thing. Now those things shouldn't be GET links (GET is supposed to be "safe" by spec), but it's a weird, wild, web out there.
2. Running analytics/ads during the speculation. So page views seem to increase, even if user never visits those pages. Many analytics providers (e.g. Google Analytics, GPT, Google AdWords) are prerender-aware, and don't log "views" until the pages is activated but not all.
3. More downloads to your user. The "danger" here is in using your users bandwidth. Yes can have a real cost depending how much your users pay for their bandwidth. Let's say your pages are 10MB then speculating 9 future pages on top of your current page could result in 100MB being downloaded by your users. It's a bit rude at least!
4. More downloads from your server. Let's say you load 9 extra pages for every 1 page of your own, then you could DDOS yourself if running with limited capacity!

We have [a number of mitigations](https://www.youtube.com/watch?v=bSua9vMdrNM&t=1319s) which reduce the cost making it cheaper that people think (primarily we reuse cached assets, and don't load third-party iframes like video players, maps, chat widgets — which are often heavy).

> How do we determine if it's over-speculating, and how do we deal with it?

I don't think there's a single right answer. Light sites (like HTML Hell!) can afford (both from a user and client perspective) to speculate more. Heavier sites there's more costs. And it depends on how often and accurately you can predict next navigations. Using the hover or mousedown eagerness settings means it more likely that speculations will be used, so it's not "wasted" cost.

> Can we take the first code snippet from your post, add it to our HTML boilerplate, and use it on every site?

I wouldn't suggest that, to be honest for above reasons.

I think it's fine for static blogs. Or `prefetch` with `eagerness: conservative`. This is what WordPress did for [ALL WordPress sites as of 6.8](https://make.wordpress.org/core/2025/03/06/speculative-loading-in-6-8/) as WordPress blogs are mostly static sites, and they added some special checks in for WooCommerce their popular ecommerce platform. Shopify did similar for [ALL their stores](https://performance.shopify.com/blogs/blog/speculation-rules-at-shopify) but after much more checks.

More complex web apps, might not want to speculate without testing for any negative impact, and excluding URLs (like log outs, add-to-cart).

So overall, no I wouldn't suggest adding to a a generic HTML boilerplate that might be used on lots of sites.

I can add some of this context to the post if you want? Though it is getting away from the intent of the post. I can also remove or tone down this note if you think it's a bit much? I was trying to make sure people realise there are things to consider.

File: hell/adventcalendar/2025/3/index.md (hunk 6 above)
Comment (tunetheweb) [reply hell/adventcalendar/2025/3/index.md]: OK I changed it to:

> One common technique is to prefetch the HTML document with `eager` value  as that  is often relatively cheap. Then upgrade this `prefetch` to a full `prerender` on `moderate` when you have more signals that the users may click on the link, and so think it's worthwhile to speculatively start to render the page in full, as prerender has more costs including downloading subresources and using memory and CPU needed to render the page.

Does that make it clearer?

> Adding to one my revious questions: is this something I would want to add to my HTML boilerplate?

As per previous answer, no I wouldn't suggest it for same reasons as previous.

File: hell/adventcalendar/2025/3/index.md (hunk 5 above)
Comment (matuzo) [reply hell/adventcalendar/2025/3/index.md]: That's super interesting, thank you!

Please don't remove the note. It would be great if you could add some of it to the post. Even if you think it's getting away from the intent of the post, I think it adds valuable information.

File: hell/adventcalendar/2025/3/index.md (hunk 6 above)
Comment (matuzo) [reply hell/adventcalendar/2025/3/index.md]: Yes, thank you.

File: hell/adventcalendar/2025/3/index.md (hunk 5 above)
Comment (tunetheweb) [reply hell/adventcalendar/2025/3/index.md]: OK I've link to my guide for more information:

> <p class="highlight"><strong>Note:</strong> Speculating has a cost. To both your users and for sites with potential increased traffic and resulting infrastructure usage. Always weigh those costs against the benefit to the user. Additionally, the are risks for more complex sites for speculating. See the <a href="https://developer.chrome.com/docs/web-platform/implementing-speculation-rules">Guide to implementing speculation rules for more complex sites</a> for more information.</p>

Think that strikes the balance between giving more info for those looking for it, without expanding the post too much.

WDYT?

Comment (matuzo) [reply hell/adventcalendar/2025/3/index.md]: Perfect, thank you!

File: hell/adventcalendar/2025/3/index.md [hunk 20]
```diff
@@ -1,35 +1,123 @@
...
+    url: "https://mastodon.social/@tunetheweb"
+    link_label: "@tunetheweb@mastodon.social"
+intro: "<p>The Speculation Rules API is becoming even better with improved mobile support, and a few other improvements to help speed up your websites.</p>"
 image: "advent25_3"
 ---
-Some text.
-Some text.
 
-Some text. Some text.
+The [Speculation Rules API](https://developer.mozilla.org/docs/Web/API/Speculation_Rules_API) allows you to speed up future navigations by prefetching or even prerendering URLS in advance of a user actually clicking a link. When the link is clicked, the speculation is used, and the user experiences a faster load than if no speculation was used.
```
Comment (hidde) [inline hell/adventcalendar/2025/3/index.md]: nitpick: would suggest “URLs” instead of  “URLS”

File: hell/adventcalendar/2025/3/index.md [hunk 21]
```diff
@@ -1,35 +1,123 @@
...
+  if (!mediaQuery.matches) {
+      document.querySelector('video').autoplay = true;
+  }
+</script>
+
+In the video, you can see that as you scroll down the page, links are being successfully prerendered, ready for the user to browse to the page.
+
+<p class="highlight"><strong>Note:</strong> Speculating has a cost. To both your users and for sites with potential increased traffic and resulting infrastructure usage. Always weigh those costs against the benefit to the user.</p>
+
+To conserve memory, Chrome keeps up to two speculations in memory at a time. As the user scrolls further and new links — which are more likely to be clicked on — enter the viewport, the old prerenders are cancelled. Should these links be re-speculated (for example, but the user scrolls back up), then they will be fetched from the HTTP cache and so prerender even faster.
```
Comment (hidde) [inline hell/adventcalendar/2025/3/index.md]: suggestion to improve readability:

```suggestion
To conserve memory, Chrome keeps up to two speculations in memory at a time. As the user scrolls further and new links — which are more likely to be clicked on — enter the viewport, the old prerenders are cancelled. In case these links are re-speculated, for example, when the user scrolls back up, they will be fetched from the HTTP cache. That way, they prerender even faster.
```

File: hell/adventcalendar/2025/3/index.md [hunk 22]
```diff
@@ -1,35 +1,123 @@
...
 
-## Heading
+## `eager` eagerness improvements
 
-Some text.
+As well as the improvements to the `moderate` eagerness, the `eager` value has also been recently changed to offer an option somewhere between `immediate` (where links are speculated as soon as possible) and `moderate`.
```
Comment (hidde) [inline hell/adventcalendar/2025/3/index.md]: Suggestion for readability:

```suggestion
The `eager` value has also been changed. It now offers an option somewhere between `immediate`, where links are speculated as soon as possible, and `moderate`.
```

File: hell/adventcalendar/2025/3/index.md [hunk 23]
```diff
@@ -1,35 +1,123 @@
...
 
+On desktop, `moderate` rules trigger after a 10 millisecond hover, and on mobile we consider ALL links in the viewport after the user has stopped scrolling for 100 milliseconds, rather than the more restrictive set of heuristics above.
```
Comment (hidde) [inline hell/adventcalendar/2025/3/index.md]: for readability: 

```suggestion
On desktop, `moderate` rules trigger after a 10 millisecond hover. On mobile, we consider _all_ links in the viewport after the user has stopped scrolling for 100 milliseconds, rather than the more restrictive set of heuristics above.
```

File: hell/adventcalendar/2025/3/index.md [hunk 24]
```diff
@@ -1,35 +1,123 @@
...
+{
+  "prerender_until_script": [{
+    "where": { "href_matches": "/*" },
+    "eagerness": "moderate"
+  }]
+}
+</script>
+```
+
+This enhancement should be released next year, but in the meantime, have a place and <a href="https://bsky.app/profile/tunetheweb.com">let me know</a> how you get on!
```
Comment (hidde) [inline hell/adventcalendar/2025/3/index.md]: Maybe my non native English, but I am not sure what you mean by “have a place”?

Comment (hidde) [reply hell/adventcalendar/2025/3/index.md]: This could be two sentences: 

```suggestion
This enhancement should be released next year. In the meantime, have a place and <a href="https://bsky.app/profile/tunetheweb.com">let me know</a> how you get on! 
```

**Review by hidde — COMMENTED** [review]
Great post Barry, learned a tonne. Added a few nitpicks and readability suggestions.

File: hell/adventcalendar/2025/3/index.md (hunk 20 above)
Comment (tunetheweb) [reply hell/adventcalendar/2025/3/index.md]: Agreed!

```suggestion
The [Speculation Rules API](https://developer.mozilla.org/docs/Web/API/Speculation_Rules_API) allows you to speed up future navigations by prefetching or even prerendering URLs in advance of a user actually clicking a link. When the link is clicked, the speculation is used, and the user experiences a faster load than if no speculation was used.
```

File: hell/adventcalendar/2025/3/index.md (hunk 24 above)
Comment (tunetheweb) [reply hell/adventcalendar/2025/3/index.md]: Oops. It's meant to be "have a play"

```suggestion
This enhancement should be released next year. In the meantime, have a play and <a href="https://bsky.app/profile/tunetheweb.com">let me know</a> how you get on! 
```

File: hell/adventcalendar/2025/3/index.md (hunk 21 above)
Comment (tunetheweb) [reply hell/adventcalendar/2025/3/index.md]: Tweaked your suggestion slightly:

```suggestion
To conserve memory, Chrome keeps up to two speculations in memory at a time. As the user scrolls further and new links — which are more likely to be clicked on — enter the viewport, the old prerenders are cancelled. These links can be re-speculated, for example, when the user scrolls back up, in which case they can be fetched from the HTTP cache. That way, they prerender even faster.
```

**Comment by tunetheweb** [comment]
Thanks for the review @hidde ! Accepted your suggestions (sometimes with a small tweak).

@matuzo , let me know if there's anything else to do for this. but think we're good now from my side.

**Comment by matuzo** [comment]
@tunetheweb Perfect, thank you!
